webtop/api.py

```diff
@@ -0,0 +1,159 @@
+        self,
+        url: str,
+        name_resolution_overrides: Optional[Dict[str, str]],
+        method: str = "GET",
+        number_of_running_requests: int = 100_000,
+        number_of_workers: int = 1,
+        timeout: int = 10,
+    ):
+        self.url = url
+        self.method = HTTPMethod[method]
+        self.__name_resolution_overrides = name_resolution_overrides or {}
+        self.__number_of_running_requests = number_of_running_requests
+        self.__number_of_workers = number_of_workers
+        self.__results: Deque[Result] = deque(maxlen=number_of_running_requests)
+        self.__stop_event = Event()
+        self.__timeout = timeout
+
+    def __session(self) -> aiohttp.ClientSession:
+        return aiohttp.ClientSession(
+            timeout=aiohttp.ClientTimeout(connect=self.__timeout),
+            connector=aiohttp.TCPConnector(
+                force_close=True, limit=0, resolver=Resolver(custom_mappings=self.__name_resolution_overrides)
+            ),
+        )
+
+    async def start(self,) -> None:
+        async with self.__session() as session:
+
+            async def worker() -> None:
+                while not self.__stop_event.is_set():
+                    result = await request(url=self.url, method=self.method.value, session=session)
+                    self.__results.append(result)
+
+            event_loop = asyncio.get_event_loop()
+            tasks = []
+            for _ in range(self.__number_of_workers):
+                tasks.append(event_loop.create_task(worker()))
+            self.__tasks = tasks
+
+    async def stop(self) -> None:
+        self.__stop_event.set()
+        asyncio.gather(*self.__tasks)
+
+    def get_results(self) -> Set[Result]:
+        return set(copy.deepcopy(self.__results))
+
+    def get_statistics(self) -> "Statistics":
+        return Statistics(self)
+
+
+class Statistics(object):
+    def __init__(self, runner: Runner):
+        self.url = runner.url
+        self.method = runner.method.value
+        results = runner.get_results()
+        self.sample_size = len(results)
+        self.reason_counts: Dict[str, int] = {}
+        number_of_successful_results = 0
+        number_of_responses = 0
+        sum_latency = 0
+
+        for result in results:
+            if result.is_success:
+                number_of_successful_results += 1
+
+            if isinstance(result, ResponseResult):
+                number_of_responses += 1
+                sum_latency += math.ceil(result.elapsed / datetime.timedelta(milliseconds=1))
+                assert result.response is not None
+                reason = f"HTTP {result.response.status}"
+            elif isinstance(result, ErrorResult):
+                reason = str(type(result.error).__name__)
+
+            self.reason_counts[reason] = self.reason_counts.get(reason, 0) + 1
+
+        self.success_rate = number_of_successful_results / number_of_responses * 100.0 if self.sample_size > 0 else 0.0
```

Comment:
This can raise a division by zero error if `number_of_responses` is 0. I don't know exactly how it gets in that state, but I've encountered it. I'd recommend this:
```python
        self.success_rate = 0.0
        if self.sample_size > 0 and number_of_responses > 0:
            self.success_rate = number_of_successful_results / number_of_responses * 100.0 
```